File: programmers/js/new/level3/가장_먼_노드.js
function solution(n, edge) {
    const visited = Array(n + 1).fill(false);
    const distance = Array(n + 1);
    const edges = Array.from(Array(n + 1), () => []);
    edge.forEach((e) => {
        const [a, b] = e;
        edges[a].push(b);
        edges[b].push(a);
    });
    distance[1] = 0;
    const q = [1];
    visited[1] = true;
    let count = 0;
    while (q.length > 0) {
        const qsize = q.length;
        count++;
        for (let i = 0; i < qsize; ++i) {
            const current = q.shift();
            for (let j = 0; j < edges[current].length; ++j) {
                const next = edges[current][j];
                if (!visited[next]) {
                    visited[next] = true;
                    distance[next] = count;
                    q.push(next);
                }
            }
        }
    }
    let maxDintanceNodes = [];
    let maxDistance = 0;
    for (let i = 1; i < distance.length; ++i) {
        if (distance[i] > maxDistance) {
            maxDistance = distance[i];
            maxDintanceNodes = [i];
        } else if (distance[i] === maxDistance) {
            maxDintanceNodes.push(i);
        }
    }
    return maxDintanceNodes.length;
}

console.log(
    solution(6, [
        [3, 6],
        [4, 3],
        [3, 2],
        [1, 3],
        [1, 2],
        [2, 4],
        [5, 2],
    ])
);

File: programmers/python/level3/가장_먼_노드.py
from collections import deque
def solution(n, edge):
    adj_list = [[] for i in range(n+1)]
    for f, t in edge:
        adj_list[f].append(t)
        adj_list[t].append(f)
    visited = [False for i in range(n+1)]
    visited[1] = True
    q = deque([1])
    while len(q) > 0:
        qsize = len(q)
        for i in range(qsize):
            num = q.popleft()
            for next in adj_list[num]:
                if visited[next] == False:
                    visited[next] = True
                    q.append(next)
    return qsize


print(solution(6, [[3, 6], [4, 3], [3, 2], [1, 3], [1, 2], [2, 4], [5, 2]]))